snuba/util.py

```diff
@@ -444,7 +444,11 @@ def raw_query(body, sql, client, timer, stats=None):
                             query_settings['max_threads'] = 1
 
                         try:
-                            result = NativeDriverReader(client).execute(
+                            result = HTTPReader(
+                                settings.CLICKHOUSE_HOST,
+                                settings.CLICKHOUSE_HTTP_PORT,
+                                {'output_format_json_quote_64bit_integers': '0'},  # this is bad, but it makes the tests pass
```

Comment:
So. [This is explained in the ClickHouse documentation](
https://clickhouse.yandex/docs/en/operations/settings/settings/#session_settings-output_format_json_quote_64bit_integers):

> If the value is true, integers appear in quotes when using JSON* Int64 and UInt64 formats (for compatibility with most JavaScript implementations); otherwise, integers are output without the quotes.

Why is this important? See https://github.com/getsentry/sentry/issues/6593#issuecomment-345386882. This has obviously been an issue for a minute with Sentry, even before Snuba. It is still an issue with Sentry. It is also an issue wherever we fetch really big numbers from Snuba and proxy the response directly to JavaScript (e.g. discover, I think?)

Note: The Python JSON parser will treat numbers as Python numeric types by default ([see `parse_int` and `parse_float` in `json.load`](https://docs.python.org/3/library/json.html#json.load)), so integers are integers (or longs), [floats are floats with 53 bits of precision](https://docs.python.org/3/tutorial/floatingpoint.html). Nothing is lost. If we were to go from `f(x: str) -> Union[int, float]` in Sentry and then serialize *that* value as JSON, we lose the benefit of returning strings from Snuba.

So, moving forward, we have a few different options:

1. Leave this as-is (but probably remove the comment.) Continue to return big integers as numbers from the API. This is the easiest option with the least change, but the least correct.
2. Remove this (or switch back to the default setting.) Return strings from the API. Fix the tests to work with strings. This is a more complicated option — not for Snuba, but for integrating with the users of Snuba. This will potentially require some coordination with the Sentry frontend and backend to ensure big numbers are casted to numeric types, when needed.
3. Leave this as-is (but probably remove the comment.) Use the schema to know which numbers need to be transformed to strings when returning the API response. Fix the tests to work with strings. This has the same integration burden as the previous option, and is also more complex for Snuba. This is probably the most flexible and long-term sustainable option (changes to the `Reader` interface are not coupled to the output format as closely and are less likely to introduce regressions), but will take the most work.
4. Make `output_format_json_quote_64bit_integers` a *Snuba* query setting, and either handle it by passing the option to ClickHouse, or transforming the numbers ourselves — basically, a query-time configurable version of the other options (we'd still have to decide between 2 and 3 when dealing with strings.)

Reply:
So just to clarify:
this way we would return a json like :

```
{
    something: 123123123123213
}
```

is that correct?

About the options I am not sure I follow why the 3rd is the long-term sustainable. Wouldn't we have a mix of integers returned with quotes and integers returned as integers ?

Reply:
> So just to clarify:
> this way we would return a json like :
> 
> ```
> {
>     something: 123123123123213
> }
> ```
> 
> is that correct?

Not really. It depends. Here is a small number (all options):

```
ted@veneno % echo 'SELECT 1 as number FORMAT JSON' | curl 'http://localhost:8123/' --data-binary @-                                                                                                             
{
	"meta":
	[
		{
			"name": "number",
			"type": "UInt8"
		}
	],

	"data":
	[
		{
			"number": 1
		}
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.0000635,
		"rows_read": 1,
		"bytes_read": 1
	}
}
```

Here is a big number with quotes (option 2, 3, sort of 4):

```
ted@veneno % echo 'SELECT 9007199254740992 as number FORMAT JSON' | curl 'http://localhost:8123/' --data-binary @-
{
	"meta":
	[
		{
			"name": "number",
			"type": "UInt64"
		}
	],

	"data":
	[
		{
			"number": "9007199254740992"
		}
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.0000545,
		"rows_read": 1,
		"bytes_read": 1
	}
}
```

Here is a big number with quoting disabled (option 1, will lose precision in some JSON parsers):

```
ted@veneno % echo 'SELECT 9007199254740992 as number FORMAT JSON' | curl 'http://localhost:8123/?output_format_json_quote_64bit_integers=0' --data-binary @-
{
	"meta":
	[
		{
			"name": "number",
			"type": "UInt64"
		}
	],

	"data":
	[
		{
			"number": 9007199254740992
		}
	],

	"rows": 1,

	"statistics":
	{
		"elapsed": 0.0000372,
		"rows_read": 1,
		"bytes_read": 1
	}
}
```
 
> About the options I am not sure I follow why the 3rd is the long-term sustainable. Wouldn't we have a mix of integers returned with quotes and integers returned as integers ?

Yes, we would have a mix (if you mixed 64-bit and non-64 bit numbers in the same response body) — but that's the only way to ensure that some runtimes don't lose precision when displaying these large numbers. Regarding sustainability, specifically: if we decide we *should* use the string representation of the number, it seems more sustainable to implement that in Snuba, rather than deferring that behavior to the ClickHouse reader (in case it changes behavior in the future, or we add support for non-ClickHouse data stores.)

Reply:
My concern about the sustainability was the mix of quoted and non quoted numbers that would be confusing for the clients. Quoting in Snuba is fine to me, but I would always quote.

Reply:
I'm also fine with that (but we'd have to implement that on the Snuba side.)

Reply:
I like option 1, even though it might not result in Javascript being displayed properly.

Because, this problem exists in sentry right now. And having a field displayed as a number from sentry and a string from snuba will confuse clients. And I feel this kind of problem should be solved at the API layer, rather than in Snuba.

Reply:
I left this as-is (numbers unquoted) for consistency with the current implementation. I still do think we should reevaluate this later, but it doesn't need to necessarily be with this change.